app/assets/javascripts/select_two.js.es6

```diff
@@ -106,24 +109,95 @@ App.Form.Select2Input = class Select2Input {
     $selectAllLink.html(html)
   }
 
+  // called on every open
   initToggleChildren(e) {
     const self = this
     $('body').on('click', '.select2-results__group', function(e){
-
-      let ids = $(this).next('ul').find('li.select2-results__option').map(function() {
-        // Trigger change.select2 should really do this, but it doesn't, so we manually set the selected nature
-        $(this).attr('aria-selected', 'true')
-        return this.id.split('-').pop()
-      }).get()
-      let selected = $(self).find(':selected').map(function(){
+      let group_ids = self.optionGroupOptionIds(self.optionsForGroup($(this)))
+      let previously_selected = self.$select.find(':selected').map(function () {
         return this.value
       }).get()
-      // select anything within the opt group and the previously selected items
-      $(self).val(selected.concat(ids))
-      $(self).trigger('change.select2')
+      // uncheck
+      if ($(this).hasClass('j-all-selected')) {
+        let now_selected = previously_selected.filter(id => !group_ids.includes(id))
+        self.$select.val(now_selected)
+        self.$select.trigger('change.select2')
+        self.optionsForGroup($(this)).each(function () {
+          $(this).attr('aria-selected', 'false')
+        })
+        $(this).removeClass('j-all-selected')
+      }
+      // check
+      else {
+        // select anything within the opt group and the previously selected items
+        self.$select.val(previously_selected.concat(group_ids))
+        self.$select.trigger('change.select2')
+
+        // Trigger change.select2 should really do this, but it doesn't, so we manually set the selected nature
+        self.optionsForGroup($(this)).each(function () {
+          $(this).attr('aria-selected', 'true')
+        })
+        // update the optgroup to reflect selected state
+        $(this).addClass('j-all-selected')
+      }
     })
   }
 
+  updateOptGroupState(e) {
+    // FIXME: the open event is firing before the drop-down is populated completely,
+    // so the optgroups can't get populated with select none when fully selected
```

Comment:
@jgoley , if you have any thoughts about how to overcome this one, I'd love some help.  Essentially the open event is firing before the DOM is fully populated, I think this would work correctly if it waited until the list was populated.